### ISSUE TITLE
Allow aspects to be top-level entities

### ISSUE THREAD

**Issue by TimLethbridge** [umple/umple#1521]
Aspects in Umple currently must be embedded in classes and operate on matching methods in that class.

It should be possible to specify a before or after statement at the top level of an Umple program. That would take an extra parameter which would be a pattern of matching classes.

When encountered, such an option would simply copy apply the statement in all matching classes at the metamodel level. This should not require any new code generation, since the existing code generation for aspects should work. All types of aspects should work include generated methods, user-defined methods; and with or without patterns.

**Comment by TimLethbridge** [comment]
A an example if I had

```
class X {
 name;
 after setName {System.out.println("Set it";)}
}
class Y {
 name;
}

```

It would set the name of only in one class.

However the proposal is that I should be able to do this:

```
class X {
 name;

}
class Y {
 name;
}

 after {*} setName {System.out.println("Set it";)}

```

Where * matches all classes. The grammar of the after clause (and before clause) would be usable at the top level, except that there would be an extra argument with a list of patterns (where * matches all).  In the above, I could also have written 

```
 after {X,Y} setName {System.out.println("Set it";)}
```

So there could be a list of classes. Patterns like {Person*} would match classes starting with Person.

**Comment by TimLethbridge** [comment]
As to making this work at the metamodel level. During parsing, the aspects need to be saved since classes may not have been encountered yet. Then after parsing they need to be applied just as though they had been done in every matching class.

Things to test:

- Make sure this works with all variants of before and after (e.g. with generated and custom methods, with patterns of methods, with arguments and so on.
- Make sure it works with mixins and mixsets
- Make sure if you have both in-class aspects and this new top-level aspects that both injections work
- Make sure all error messages with aspects work correctly when they are top-level
- Make sure that all kinds of patterns work (matching classes)
- Make sure that any error messages about aspects point to the correct line of code

**Comment by WeiqiMaster** [comment]
For the {*} notation, if a class only contains one of the methods mentioned in the aspect statement, should that method be injected into that matching class?

**Comment by TimLethbridge** [comment]
> For the {*} notation, if a class only contains one of the methods mentioned in the aspect statement, should that method be injected into that matching class?

@WeiqiMaster I am not sure what you are asking. Please give a concrete example. The * means to match anything and applies to both class name matching (in this new feature) and method name matching.

So the following

```
after {*Color, *Paint* } convert* {System.out.println("Called a convert method";)}
```

would match any method starting with 'convert' in any class whose name ends with Color and any class with the word Paint in it's name. However if there wasn't a convert* method in a matching class, nothing would happen in that class. And likewise nothing would happen of there are no matching classes or no matching methods.

Essentially the new feature is simply supposed to find all matching classes and behave exactly as though the injection had been present in these classes (except, don't report errors if nothing matches in a given class)

**Comment by WeiqiMaster** [comment]
For example: 

after {*} setName,testFunction { doSomething(); }  
class Student
{
  name;
}
class Teacher
{
  name;
}
class Employer
{
  age;
  String testFunction() {
    return "This is a test function";
  }
}

In the above code, Class Student and Teacher only contain setName() method, and class Employer only contains testFunction() method. Will this aspect be valid?

**Comment by TimLethbridge** [comment]
@WeiqiMaster Yes,  the above code is valid.

**Comment by WeiqiMaster** [comment]
Is the asterisk allowed in the middle of a class name? Like {Color*Color}

**Comment by TimLethbridge** [comment]
@WeiqiMaster Of course an asterisk can appear in the middle. This style of patterns is called 'Glob'. The following link shows some code for converting Glob to regex that you can then use to match class names:

https://stackoverflow.com/questions/1247772/is-there-an-equivalent-of-java-util-regex-for-glob-type-patterns

**Comment by TimLethbridge** [comment]
Fixed with PR #1548

**Comment by TimLethbridge** [comment]
Now just needs user manual page before closing
